Sources/Moho/ScopeCheck.swift

```diff
@@ -0,0 +1,576 @@
+          return .lambda(next, acc)
+        }
+      }
+    case let syntax as ApplicationExprSyntax:
+      guard syntax.exprs.count > 1 else {
+        return self.scopeCheckExpr(syntax.exprs[0])
+      }
+
+      guard let headExpr = syntax.exprs[0] as? NamedBasicExprSyntax else {
+        fatalError("Cannot yet handle this case")
+      }
+
+      let n = QualifiedName(ast: headExpr.name)
+      guard n.string != "->" else {
+        assert(syntax.exprs.count == 3)
+        return .function(self.scopeCheckExpr(syntax.exprs[1]),
+                         self.scopeCheckExpr(syntax.exprs[2]))
+      }
+
+      var args = [Expr]()
+      for i in 1..<syntax.exprs.count {
+        let e = syntax.exprs[i]
+        let elim = self.scopeCheckExpr(e)
+        args.append(elim)
+      }
+
+      let head: ApplyHead
+      if self.isBoundVariable(n.name) {
+        head = ApplyHead.variable(n.name)
+      } else if let (fqn, _) = self.lookupLocalName(n.name) {
+        head = ApplyHead.definition(fqn)
+      } else if self.lookupFullyQualifiedName(n) != nil {
+        head = ApplyHead.definition(n)
+      } else {
+        // If it's not a definition or a local variable, it's undefined.
+        // Recover by introducing a local variable binding anyways.
+        self.engine.diagnose(.undeclaredIdentifier(n), node: n.node)
+        head = ApplyHead.variable(n.name)
+      }
+
+      return .apply(head, args.map(Elimination.apply))
+    case let syntax as QuantifiedExprSyntax:
+      return self.underScope { _ in
+        let telescope = syntax.bindingList.map(self.scopeCheckTelescope)
+        var type = self.scopeCheckExpr(syntax.outputExpr)
+        for (names, expr) in telescope {
+          for name in names {
+            type = Expr.pi(name, expr, type)
+          }
+        }
+        return type
+      }
+    case let syntax as TypedParameterArrowExprSyntax:
+      let telescope = syntax.parameters.map(self.scopeCheckTelescope)
+      var type = self.scopeCheckExpr(syntax.outputExpr)
+      for (names, expr) in telescope {
+        for name in names {
+          type = Expr.pi(name, expr, type)
+        }
+      }
+      return type
+    default:
+      print(type(of: syntax))
+      fatalError()
+    }
+  }
+
+  private func scopeCheckBindingList(
+    _ syntax: BindingListSyntax) -> [([Name], Expr)] {
+    var bs = [([Name], Expr)]()
+    for i in 0..<syntax.count {
+      let binding = syntax[i]
+      switch binding {
+      case let binding as NamedBindingSyntax:
+        let name = QualifiedName(ast: binding.name).name
+        guard let bindName = self.bindVariable(named: name) else {
+          fatalError()
+        }
+        bs.append(([bindName], .meta))
+      case let binding as TypedBindingSyntax:
+        bs.append(self.scopeCheckTelescope(binding.parameter))
+      default:
+        fatalError()
+      }
+    }
+    return bs
+  }
+
+  private func scopeCheckTelescope(
```

Comment:
I'll be honest, I've never quite understood the `telescope` concept in this context. A comment would be much appreciated.